Computer graphics demo fails on pow

```
clang-9: /home/blide/gsoc/src/tools/clad/lib/Differentiator/DerivativeBuilder.cpp:1456: clad::StmtDiff
clad::ForwardModeVisitor::VisitCallExpr(const clang::CallExpr*): Assertion `(CE->getNumArgs() <= 1) &&
"forward differentiation of multi-arg calls is currently broken"' failed.
```
pow is also broken in reverse mode.

```c++
double f(double x) {
  return pow(x, 2);
}
clad::gradient(f);
```
Produces
```
1.	<eof> parser at end of file
2.	LLVM IR generation of declaration 'f_grad'
3.	Generating code for declaration 'f_grad'
4.	LLVM IR generation of compound statement ('{}')
 #0 0x000055bdcd475819 llvm::sys::PrintStackTrace(llvm::raw_ostream&) /llvm-project/llvm/lib/Support/Unix/Signals.inc:533:22
 #1 0x000055bdcd4758b0 PrintStackTraceSignalHandler(void*) /llvm-project/llvm/lib/Support/Unix/Signals.inc:594:1
 #2 0x000055bdcd47374a llvm::sys::RunSignalHandlers() /llvm-project/llvm/lib/Support/Signals.cpp:68:20
 #3 0x000055bdcd4751a3 SignalHandler(int) /llvm-project/llvm/lib/Support/Unix/Signals.inc:385:1
 #4 0x00007fc52f7953c0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x153c0)
 #5 0x00007fc52dc9518b raise /build/glibc-YYA7BZ/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #6 0x00007fc52dc74859 abort /build/glibc-YYA7BZ/glibc-2.31/stdlib/abort.c:81:7
 #7 0x00007fc52dc74729 get_sysdep_segment_value /build/glibc-YYA7BZ/glibc-2.31/intl/loadmsgcat.c:509:8
 #8 0x00007fc52dc74729 _nl_load_domain /build/glibc-YYA7BZ/glibc-2.31/intl/loadmsgcat.c:970:34
 #9 0x00007fc52dc85f36 (/lib/x86_64-linux-gnu/libc.so.6+0x36f36)
#10 0x000055bdccb84b7f llvm::StoreInst::AssertOK() /llvm-project/llvm/lib/IR/Instructions.cpp:1358:3
#11 0x000055bdccb84e91 llvm::StoreInst::StoreInst(llvm::Value*, llvm::Value*, bool, unsigned int, llvm::AtomicOrdering, unsigned char, llvm::Instruction*) /llvm-project/llvm/lib/IR/Instructions.cpp:1403:1
#12 0x000055bdccb84d2e llvm::StoreInst::StoreInst(llvm::Value*, llvm::Value*, bool, unsigned int, llvm::Instruction*) /llvm-project/llvm/lib/IR/Instructions.cpp:1382:48
#13 0x000055bdccb84c92 llvm::StoreInst::StoreInst(llvm::Value*, llvm::Value*, bool, llvm::Instruction*) /llvm-project/llvm/lib/IR/Instructions.cpp:1373:68
#14 0x000055bdcd84e0e2 llvm::IRBuilder<llvm::ConstantFolder, clang::CodeGen::CGBuilderInserter>::CreateStore(llvm::Value*, llvm::Value*, bool) /llvm-project/llvm/include/llvm/IR/IRBuilder.h:1606:54
#15 0x000055bdcd84e05d llvm::IRBuilder<llvm::ConstantFolder, clang::CodeGen::CGBuilderInserter>::CreateAlignedStore(llvm::Value*, llvm::Value*, unsigned int, bool) /llvm-project/llvm/include/llvm/IR/IRBuilder.h:1651:16
#16 0x000055bdcd84ce46 clang::CodeGen::CGBuilderTy::CreateStore(llvm::Value*, clang::CodeGen::Address, bool) /llvm-project/clang/lib/CodeGen/CGBuilder.h:111:3
#17 0x000055bdcdbbccc8 clang::CodeGen::CodeGenFunction::EmitStoreOfScalar(llvm::Value*, clang::CodeGen::Address, bool, clang::QualType, clang::CodeGen::LValueBaseInfo, clang::CodeGen::TBAAAccessInfo, bool, bool) /llvm-project/clang/lib/CodeGen/CGExpr.cpp:1752:47
#18 0x000055bdcdbbceb7 clang::CodeGen::CodeGenFunction::EmitStoreOfScalar(llvm::Value*, clang::CodeGen::LValue, bool) /llvm-project/clang/lib/CodeGen/CGExpr.cpp:1765:20
#19 0x000055bdcdbbe4b6 clang::CodeGen::CodeGenFunction::EmitStoreThroughLValue(clang::CodeGen::RValue, clang::CodeGen::LValue, bool) /llvm-project/clang/lib/CodeGen/CGExpr.cpp:2020:20
#20 0x000055bdcdb8ba9b clang::CodeGen::CodeGenFunction::EmitScalarInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) /llvm-project/clang/lib/CodeGen/CGDecl.cpp:754:27
#21 0x000055bdcdb913d4 clang::CodeGen::CodeGenFunction::EmitExprAsInit(clang::Expr const*, clang::ValueDecl const*, clang::CodeGen::LValue, bool) /llvm-project/clang/lib/CodeGen/CGDecl.cpp:1886:19
#22 0x000055bdcdb91066 clang::CodeGen::CodeGenFunction::EmitAutoVarInit(clang::CodeGen::CodeGenFunction::AutoVarEmission const&) /llvm-project/clang/lib/CodeGen/CGDecl.cpp:1845:26
#23 0x000055bdcdb8e2f7 clang::CodeGen::CodeGenFunction::EmitAutoVarDecl(clang::VarDecl const&) /llvm-project/clang/lib/CodeGen/CGDecl.cpp:1286:22
#24 0x000055bdcdb890d4 clang::CodeGen::CodeGenFunction::EmitVarDecl(clang::VarDecl const&) /llvm-project/clang/lib/CodeGen/CGDecl.cpp:192:27
#25 0x000055bdcdb88dd7 clang::CodeGen::CodeGenFunction::EmitDecl(clang::Decl const&) (.localalias) /llvm-project/clang/lib/CodeGen/CGDecl.cpp:137:47
#26 0x000055bdcd862344 clang::CodeGen::CodeGenFunction::EmitDeclStmt(clang::DeclStmt const&) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:1132:3
#27 0x000055bdcd85e5ea clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:350:75
#28 0x000055bdcd85d96b clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:49:3
#29 0x000055bdcd85eb10 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:390:3
#30 0x000055bdcd85e7ec clang::CodeGen::CodeGenFunction::EmitCompoundStmt(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:376:58
#31 0x000055bdcd85e5c3 clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:349:72
#32 0x000055bdcd85d96b clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:49:3
#33 0x000055bdcd85f5c8 clang::CodeGen::CodeGenFunction::EmitLabelStmt(clang::LabelStmt const&) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:583:1
#34 0x000055bdcd85e60d clang::CodeGen::CodeGenFunction::EmitSimpleStmt(clang::Stmt const*) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:351:75
#35 0x000055bdcd85d96b clang::CodeGen::CodeGenFunction::EmitStmt(clang::Stmt const*, llvm::ArrayRef<clang::Attr const*>) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:49:3
#36 0x000055bdcd85eb10 clang::CodeGen::CodeGenFunction::EmitCompoundStmtWithoutScope(clang::CompoundStmt const&, bool, clang::CodeGen::AggValueSlot) /llvm-project/clang/lib/CodeGen/CGStmt.cpp:390:3
#37 0x000055bdcd8e5c69 clang::CodeGen::CodeGenFunction::EmitFunctionBody(clang::Stmt const*) /llvm-project/clang/lib/CodeGen/CodeGenFunction.cpp:1033:36
#38 0x000055bdcd8e68eb clang::CodeGen::CodeGenFunction::GenerateCode(clang::GlobalDecl, llvm::Function*, clang::CodeGen::CGFunctionInfo const&) /llvm-project/clang/lib/CodeGen/CodeGenFunction.cpp:1198:21
#39 0x000055bdcd90ef89 clang::CodeGen::CodeGenModule::EmitGlobalFunctionDefinition(clang::GlobalDecl, llvm::GlobalValue*) /llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:4320:3
#40 0x000055bdcd90822b clang::CodeGen::CodeGenModule::EmitGlobalDefinition(clang::GlobalDecl, llvm::GlobalValue*) /llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:2759:47
#41 0x000055bdcd9071e0 clang::CodeGen::CodeGenModule::EmitGlobal(clang::GlobalDecl) /llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:2512:5
#42 0x000055bdcd912fd4 clang::CodeGen::CodeGenModule::EmitTopLevelDecl(clang::Decl*) (.localalias) /llvm-project/clang/lib/CodeGen/CodeGenModule.cpp:5128:37
#43 0x000055bdce7b1f35 (anonymous namespace)::CodeGeneratorImpl::HandleTopLevelDecl(clang::DeclGroupRef) /llvm-project/clang/lib/CodeGen/ModuleBuilder.cpp:161:7
#44 0x000055bdce7aad48 clang::BackendConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /llvm-project/clang/lib/CodeGen/CodeGenAction.cpp:172:7
#45 0x000055bdcdf39700 clang::MultiplexConsumer::HandleTopLevelDecl(clang::DeclGroupRef) /llvm-project/clang/lib/Frontend/MultiplexConsumer.cpp:271:25
#46 0x00007fc52da8e055 clad::plugin::CladPlugin::ProcessDiffRequest(clad::DiffRequest&) (.localalias) /home/blide/gsoc/src/tools/clad/tools/ClangPlugin.cpp:152:9
clang-9: error: unable to execute command: Aborted (core dumped)
clang-9: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 9.0.1 (https://github.com/llvm/llvm-project.git c1a0a213378a458fbea1a5c77b315c7dce08fd05)
```

@blide I am able to reproduce the forward mode issue. But in the reserve mode I am getting:
```
/usr/bin/ld: /tmp/clad-eg-7a3277.o: in function `f_grad(double, double*)':
clad-eg.cpp:(.text+0x88): undefined reference to `std::pow_grad(double, int, double*)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
when I try to compile:
```
#include "clad/Differentiator/Differentiator.h"
#include <iostream>

double f(double x) {
  return pow(x, 2);
}
int main() {
  clad::gradient(f);
}
```
Your trace doesn't seem to indicate the same issue. Is there somewhere I am going wrong?

--- a/include/clad/Differentiator/DerivativeBuilder.h
+++ b/include/clad/Differentiator/DerivativeBuilder.h
@@ -66,7 +66,7 @@ namespace clad {
 
     clang::Expr* findOverloadedDefinition(clang::DeclarationNameInfo DNI,
                             llvm::SmallVectorImpl<clang::Expr*>& CallArgs);
-    bool overloadExists(clang::Expr* UnresolvedLookup,
+    bool noOverloadExists(clang::Expr* UnresolvedLookup,
                             llvm::MutableArrayRef<clang::Expr*> ARargs);
     /// Shorthand to issues a warning or error.
     template <std::size_t N>

--- a/lib/Differentiator/DerivativeBuilder.cpp
+++ b/lib/Differentiator/DerivativeBuilder.cpp
@@ -1371,7 +1371,7 @@ namespace clad {
   // This method is derived from the source code of both
   // buildOverloadedCallSet() in SemaOverload.cpp
   // and ActOnCallExpr() in SemaExpr.cpp.
-  bool DerivativeBuilder::overloadExists(Expr* UnresolvedLookup,
+  bool DerivativeBuilder::noOverloadExists(Expr* UnresolvedLookup,
                                          llvm::MutableArrayRef<Expr*> ARargs) {
     if (UnresolvedLookup->getType() == m_Context.OverloadTy) {
       OverloadExpr::FindResult find = OverloadExpr::find(UnresolvedLookup);
@@ -1435,7 +1435,7 @@ namespace clad {
       SourceLocation Loc;
       Scope* S = m_Sema.getScopeForContext(m_Sema.CurContext);
 
-      if (overloadExists(UnresolvedLookup, MARargs)) {
+      if (noOverloadExists(UnresolvedLookup, MARargs)) {
         return 0;
       }
 

--- a/lib/Differentiator/DerivativeBuilder.cpp
+++ b/lib/Differentiator/DerivativeBuilder.cpp
@@ -1456,9 +1456,7 @@ namespace clad {
     std::string s = std::to_string(m_DerivativeOrder);
     if (m_DerivativeOrder == 1)
       s = "";
-    // FIXME: add gradient-vector products to fix that.
-    assert((CE->getNumArgs() <= 1) &&
-           "forward differentiation of multi-arg calls is currently broken");
+
     IdentifierInfo* II = &m_Context.Idents.get(FD->getNameAsString() + "_d" +
                                                s + "arg0");
     DeclarationName name(II);
@@ -1486,6 +1484,11 @@ namespace clad {
 
     // Try to find an overloaded derivative in 'custom_derivatives'
     Expr* callDiff = m_Builder.findOverloadedDefinition(DNInfo, CallArgs);
+
+    // FIXME: add gradient-vector products to fix that.
+    if(!callDiff)
+      assert((CE->getNumArgs() <= 1) &&
+             "forward differentiation of multi-arg calls is currently broken");
 
     // Check if it is a recursive call.
     if (!callDiff && (FD == m_Function)) {
@@ -3229,4 +3232,4 @@ namespace clad {
 
     return result;
   }
-} // end namespace clad
+}// end namespace clad

--- a/test/Misc/RunDemos.C
+++ b/test/Misc/RunDemos.C
@@ -2,6 +2,7 @@
 // RUN: %cladclang %S/../../demos/ControlFlow.cpp -I%S/../../include 2>&1
 // RUN: %cladclang %S/../../demos/DebuggingClad.cpp -I%S/../../include 2>&1
 // RUN: %cladclang %S/../../demos/RosenbrockFunction.cpp -I%S/../../include 2>&1
+// RUN: %cladclang %S/../../demos/ComputerGraphics/SmallPT.cpp -I%S/../../include 2>&1
 
 
 //-----------------------------------------------------------------------------/

--- a/test/Misc/RunDemos.C
+++ b/test/Misc/RunDemos.C
@@ -2,7 +2,7 @@
 // RUN: %cladclang %S/../../demos/ControlFlow.cpp -I%S/../../include 2>&1
 // RUN: %cladclang %S/../../demos/DebuggingClad.cpp -I%S/../../include 2>&1
 // RUN: %cladclang %S/../../demos/RosenbrockFunction.cpp -I%S/../../include 2>&1
-// RUN: %cladclang %S/../../demos/ComputerGraphics/SmallPT.cpp -I%S/../../include 2>&1
+// RUN: %cladclang -lstdc++ -lm %S/../../demos/ComputerGraphics/SmallPT.cpp -I%S/../../include 2>&1
 
 
 //-----------------------------------------------------------------------------/

